tools/device-farm-runner/run_on_aws_devicefarm.py

```diff
@@ -0,0 +1,336 @@
+#!/usr/bin/env python3
+
+import datetime
+import logging
+import os
+import random
+import string
+import sys
+import time
+from enum import Enum
+from logging import info
+from typing import Any, Dict, List, Optional
+from warnings import warn
+
+import boto3
+import requests
+
+
+POLLING_DELAY_IN_SECOND = 5
+MAX_UPLOAD_WAIT_IN_SECOND = 600
+
+# NB: This is the curated top devices from AWS. We could create our own device
+# pool if we want to
+DEFAULT_DEVICE_POOL_ARN = (
+    "arn:aws:devicefarm:us-west-2::devicepool:082d10e5-d7d7-48a5-ba5c-b33d66efa1f5"
+)
+
+
+# Device Farm report type
+class ReportType(Enum):
+    RUN = "run"
+    JOB = "job"
+    SUITE = "suite"
+    TEST = "test"
+
+
+DEVICE_FARM_BUCKET = "gha-artifacts"
+
+logging.basicConfig(level=logging.INFO)
+
+
+def parse_args() -> Any:
+    from argparse import ArgumentParser
+
+    parser = ArgumentParser("Run iOS tests on AWS Device Farm")
+    parser.add_argument(
+        "--project-arn", type=str, required=True, help="the ARN of the project on AWS"
+    )
+    parser.add_argument(
+        "--app-file", type=str, required=True, help="the iOS ipa app archive"
+    )
+    parser.add_argument(
+        "--xctest-file",
+        type=str,
+        required=True,
+        help="the XCTest suite to run",
+    )
+    parser.add_argument(
+        "--name-prefix",
+        type=str,
+        required=True,
+        help="the name prefix of this test run",
+    )
+    parser.add_argument(
+        "--device-pool-arn",
+        type=str,
+        default=DEFAULT_DEVICE_POOL_ARN,
+        help="the name of the device pool to test on",
+    )
+    parser.add_argument(
+        "--workflow-id",
+        type=str,
+        default="invalid-workflow-id",
+        help="the workflow run ID",
+    )
+    parser.add_argument(
+        "--workflow-attempt",
+        type=int,
+        default=0,
+        help="the workflow run attempt",
+    )
+
+    return parser.parse_args()
+
+
+def upload_file(
+    client: Any,
+    project_arn: str,
+    prefix: str,
+    filename: str,
+    filetype: str,
+    mime: str = "application/octet-stream",
+) -> str:
+    """
+    Upload the app file and XCTest suite to AWS
+    """
+    r = client.create_upload(
+        projectArn=project_arn,
+        name=f"{prefix}_{os.path.basename(filename)}",
+        type=filetype,
+        contentType=mime,
+    )
+    upload_name = r["upload"]["name"]
+    upload_arn: str = r["upload"]["arn"]
+    upload_url = r["upload"]["url"]
+
+    with open(filename, "rb") as file_stream:
+        info(f"Uploading {filename} to Device Farm as {upload_name}...")
+        r = requests.put(upload_url, data=file_stream, headers={"content-type": mime})
+        if not r.ok:
+            raise Exception(f"Couldn't upload {filename}: {r.reason}")
+
+    start_time = datetime.datetime.now()
+    # Polling AWS till the uploaded file is ready
+    while True:
```

Comment:
`while True` is kind of movetón :)
```suggestion
    while status != "SUCCESS":
```